### PR TITLE
Feature/st 175 translations

### PR DESCRIPTION
**Pull request by james-simpson-ST**
Translation of labels, field messages and notification frame messages added

Still need to add test cases and remaining translations but framework is in place

### CODE REVIEW

File: src/core/shared/FormField.ts [hunk 1]
```diff
@@ -2,28 +2,51 @@ import Formatter from './Formatter';
...
+    this._translator = new Translator(this._params.locale);
+    this.setLabelText();
+  }
+
+  protected setLabelText() {
+    this._labelElement.innerHTML = this._translator.translate(this.getLabel());
+  }
+
+  protected getLabel() {
+    return '';
```
Comment (pgs-gmartowski) [inline src/core/shared/FormField.ts]: it looks strange, is it already finished ?

File: src/core/shared/FormField.ts [hunk 2]
```diff
@@ -137,7 +165,7 @@ export default class FormField extends Frame {
    * @param messageText
    */
   setMessage(messageText: string) {
```
Comment (pgs-gmartowski) [inline src/core/shared/FormField.ts]: function should be private, protected or public

File: src/core/shared/FormField.ts [hunk 3]
```diff
@@ -137,7 +165,7 @@ export default class FormField extends Frame {
...
-    this._messageElement.innerText = messageText;
+    this._messageElement.innerText = this._translator.translate(messageText);
   }
 
   setValue(value: string) {
```
Comment (pgs-gmartowski) [inline src/core/shared/FormField.ts]: function should be private, protected or public

File: src/core/shared/Translator.ts [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+      debug: false,
+      resources: {
+        de_DE: { translation: de },
+        en_EN: { translation: en },
+        fr_FR: { translation: fr }
+      }
+    });
+  }
+
+  public translate(text: string) {
```
Comment (pgs-gmartowski) [inline src/core/shared/Translator.ts]: nth: we can convert it to:
```
 public translate = (text: string) => i18next.t(text);
```

**Comment by pgs-gmartowski** [comment]
> Translation of labels, field messages and notification frame messages added
> 
> Still need to add test cases and remaining translations but framework is in place

@james-simpson-ST Are tasks 153, 154 and 175 included in this PR ?

File: src/core/shared/Translator.ts (hunk 4 above)
Comment (james-simpson-ST) [reply src/core/shared/Translator.ts]: Done

File: src/core/shared/FormField.ts (hunk 3 above)
Comment (james-simpson-ST) [reply src/core/shared/FormField.ts]: Done

File: src/core/shared/FormField.ts (hunk 2 above)
Comment (james-simpson-ST) [reply src/core/shared/FormField.ts]: Done

File: src/core/shared/FormField.ts (hunk 1 above)
Comment (james-simpson-ST) [reply src/core/shared/FormField.ts]: Raised Error("not implemented") is this the best approach for a parent class that needs the function defined in the child classes in javascript/typescript?

File: src/components/security-code/SecurityCode.ts [hunk 5]
```diff
@@ -55,6 +56,10 @@ export default class SecurityCode extends FormField {
     this.sendState();
   }
 
+  protected getLabel() {
```
[the remark this answers is not in the document]
Comment (james-simpson-ST) [reply src/components/security-code/SecurityCode.ts]: Done

File: src/components/expiration-date/ExpirationDate.ts [hunk 6]
```diff
@@ -56,4 +57,8 @@ export default class ExpirationDate extends FormField {
     let dataFormatted = Formatter.maskExpirationDate(data);
     this.setValue(dataFormatted);
   }
+
+  protected getLabel() {
```
[the remark this answers is not in the document]
Comment (james-simpson-ST) [reply src/components/expiration-date/ExpirationDate.ts]: Done

File: src/components/card-number/CardNumber.ts [hunk 7]
```diff
@@ -81,6 +81,10 @@ export default class CardNumber extends FormField {
     this.sendState();
   }
 
+  protected getLabel() {
+    return Language.translations.LABEL_CARD_NUMBER;
+  }
```
[the remark this answers is not in the document]
Comment (james-simpson-ST) [reply src/components/card-number/CardNumber.ts]: Done

**Comment by james-simpson-ST** [comment]
> @james-simpson-ST Are tasks 153, 154 and 175 included in this PR ?

Yes they became trivial once the translation mechanism was added so I have done them all on one task

Still need to perform self review and write unit tests for functionality

File: src/components/security-code/SecurityCode.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (james-simpson-ST) [reply src/components/security-code/SecurityCode.ts]: Had to revert out of this change after getLabel started to throw not implemented error to allow function to be overridden by child

File: src/components/expiration-date/ExpirationDate.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (james-simpson-ST) [reply src/components/expiration-date/ExpirationDate.ts]: Had to revert out of this change after getLabel started to throw not implemented error to allow function to be overridden by child

File: src/components/card-number/CardNumber.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (james-simpson-ST) [reply src/components/card-number/CardNumber.ts]: Had to revert out of this change after getLabel started to throw not implemented error to allow function to be overridden by child

**Comment by james-simpson-ST** [comment]
This now has unit tests and can be reviewed

**Comment by james-simpson-ST** [comment]
N.B. we need the remaining translation languages and final real translations (I will organise this with our translation partner)

File: src/core/shared/FormField.ts (hunk 1 above)
Comment (pgs-gmartowski) [reply src/core/shared/FormField.ts]: oh, now I see ...
that's why I sometimes hate typescript - it forces inheritance in JavaScript which is very bad approach.
if tslint raises such error we can omit it in such cases.

File: src/core/shared/FormField.ts [hunk 8]
```diff
@@ -1,29 +1,51 @@
...
   }
 
+  public onInit() {
+    super.onInit();
+    this._translator = new Translator(this._params.locale);
+    this.setLabelText();
+  }
+
+  public getLabel(): string {
+    throw new Error('Method not implemented.');
```
Comment (pgs-gmartowski) [inline src/core/shared/FormField.ts]: IMO this text should be in Language class.

Comment (james-simpson-ST) [reply src/core/shared/FormField.ts]: Good point - Done

**Pull request merged**